rfcs/0062-content-addressed-paths.md

```diff
@@ -0,0 +1,286 @@
+    have here `dynamic(contentAddressed.drv) == contentAddressed.drv`
+  - We realise `dynamic(contentAddressed.drv)`. This gives us an output path
+    `out(dynamic(contentAddressed.drv))`
+  - We compute `ca(contentAddressed.drv)` and notice that the
+    path already exists (since it's the same as the one we built previously),
+    so we discard the result.
+- We build `dependent.drv`
+  - We first compute `dynamic(dependent.drv)` to replace its
+    inputs by their real output path.
+    In that case, we replace `contentAddressed.drv!out` by
+    `ca(contentAddressed.drv)`
+  - We notice that `dynamic(dependent.drv)` is the same as before (since
+    `ca(contentAddressed.drv)` is the same as before), so we
+    just return the already existing path
+- We build `transitivelyDependent.drv`
+  - We first compute `dynamic(transitivelyDependent.drv)` to replace its
+    inputs by their real output path.
+    In that case, that means replacing `dependent.drv!out` by
+    `out(dynamic(dependent.drv))`
+  - Here again, we notice that `dynamic(transitivelyDependent.drv)` is the same as before,
+    so we don't build anything
+
+## Nix-build process
+
+### Aliases paths
+
+To allow this, we add a new type of store path: aliases paths.
+These paths don't actually exist in the store, just in the database and point to
+another path (so they are morally symlinks, but inside the db rather than
+on-disk)
+
+### Building a ca derivation
+
+ca derivations are derivations with the `__contentAddressed` argument set to
+`true`.
+
+The process for building a content-adressed derivation is the following:
+
+- We build it like a normal derivation to get an output path `$out`.
+- We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
+- We move `$out` to `/nix/store/$chash-$name`
+- We create an alias path from `$out` to `/nix/store/$chash-$name`
+
+[^modulo-hashing]:
+
+  We can possibly normalize all the self-references before
+  computing the hash and rewrite them when moving the path to handle paths with
+  self-references, but this isn't strictly required for a first iteration
+
+### Building a normal derivation
+
+The process for building a normal derivation is the following:
+
+- We look into the drv for all the inputs paths of the build
+- For each input path, we look whether the path is an alias. If so we replace it
+  by its target
+- We compute the `dynamic` output of the derivation from the patched version
+- We then try to substitute and build the new derivation
+- We create an alias path from the `static` output to the `dynamic` one
+
+## Wrapping it up
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Obviously, this makes the Nix model more complicated than what it is now. In
+  particular, the caching model needs some modifications (see [caching]);
+
+- We specify that only a sub-category of derivations can safely be marked as
+  `contentAddressed`, but there's no way to enforce these restricitions;
```

Comment:
It's a bigger problem than it might look like, as it means that trivial updates can break the CA marking for reasons not worth mentioning in the upstream changelog.

Reply:
Definitely :)

Maybe that could be clearly stated, but the original scope of this work was to be able to mark very specific derivations that were clearly guaranteed to be deterministic, in which case the problem was less important

Reply:
Well, the question «why not just propagate CA» shows that writing more is a good idea.

I do think that stressing the limitations in a few key places is also a nice thing to do (people should be able to apply RFC as passed, not what was intended and not what was discussed, after all… we should not treat ourselves worse than we treat computers!)